Update Compiler.php

owsPro opened the pull request:
line 22: private $lastLine; // It can be safely removed, see also line 52 & 175

line 51 $this->lastLine = null; // deletet line

line 175: if ($node->getTemplateLine() != $this->lastLine) { // set it from lastLine to sourceLine

line 175: if ($node->getTemplateLine() !== $this->lastLine) { // conditional can be simplified for strict comparison

line 189: return $this->debugInfo ?? []; // handle the case when no template lines have been added and returning an empty array in that case

insert line to 209: $this->indentation ??= 0;  // check to prevent the indentation from becoming negative

Commit:
Update Compiler.php

line 22: private $lastLine; // It can be safely removed, see also line 52 & 175

line 51 $this->lastLine = null; // deletet line

line 175: if ($node->getTemplateLine() != $this->lastLine) { // set it from lastLine to sourceLine

line 175: if ($node->getTemplateLine() !== $this->lastLine) { // conditional can be simplified for strict comparison

line 189: return $this->debugInfo ?? []; // handle the case when no template lines have been added and returning an empty array in that case

insert line to 209: $this->indentation ??= 0;  // check to prevent the indentation from becoming negative

Commit:
Update Compiler.php

Commit:
Update Compiler.php

Commit:
Update Compiler.php

Use now the constructor property promotion

GromNaN commented on src/Compiler.php:
```diff
@@ -48,16 +45,13 @@ public function getSource(): string
         $this->indentation = $indentation;
         $this->varNameSalt = 0;
-
-        return $this;
```
This is a breaking change. The phpdoc tell the returned value.

GromNaN commented on src/Compiler.php:
```diff
@@ -186,7 +180,7 @@ public function getDebugInfo(): array
         ksort($this->debugInfo);
 
-        return $this->debugInfo;
+        return $this->debugInfo ?? [];
```
If debugInfo is not an array, the call to ksort would fail.

owsPro replied:
Do we have data in debuginfo that is not available as an array, although the function :: requires an array ?

GromNaN replied:
debugInfo is always an array.

GromNaN commented on src/Compiler.php:
```diff
@@ -206,6 +200,7 @@ public function indent(int $step = 1)
      */
     public function outdent(int $step = 1)
     {
+        $this->indentation ??= 0;
```
Not necessary, the comparison with null is fine.

owsPro replied:
As it was originally set with step=1, 0 is also set as an error. Should this be explicit?

GromNaN reviewed (changes requested):
Did you use any static analysis tool to detect this changes? I'm not sure they are correct, except the property promotion which is syntactic sugar.

Maybe add a test if you found edge cases.

Commit:
Update Environment.php

Some additions, as well as a cache for compiled sources in the compileSource function.

GromNaN commented on src/Compiler.php:
```diff
@@ -169,7 +163,7 @@ public function repr($value)
     public function addDebugInfo(Node $node)
     {
-        if ($node->getTemplateLine() != $this->lastLine) {
+        if ($node->getTemplateLine() !== $this->sourceLine) {
```
`$this->lastLine` is still used line 173.

owsPro replied:
rewritten it as follows:

public function addDebugInfo(Node $node)
{
    	$this->sourceLine = $node->getTemplateLine();

    	$this->write("// line {$node->getTemplateLine()}\n");

    	$templateLines = substr_count($node->getSource(), PHP_EOL);
    	$this->sourceLine += $templateLines;
    	$this->sourceOffset = strlen($this->source);

    	$this->source .= $node->getSource();

    	$this->debugInfo[$node->getTemplateLine()] = $node->getTemplateLine();

    	return $this;
}

GromNaN commented on src/Environment.php:
```diff
@@ -512,19 +512,33 @@ public function compile(Node $node): string
-            throw new SyntaxError(sprintf('An exception has been thrown during the compilation of a template ("%s").', $e->getMessage()), -1, $source, $e);
-        }
-    }
+	public function compileSource(Source $source): string
```
Indenting with tabs is incorrect.

owsPro replied:
I have rewritten it as follows:

public function compileSource(Source $source): string
{
    	$cacheKey = $this->getCacheKey($source->getName());
    	$content = $this->cache->get($cacheKey);

    	if ($content === null) {
        	$content = $this->compile($this->parse($this->tokenize($source)));
        	$this->cache->put($cacheKey, $content);
    	}

    	return $content;
}

public function tokenize(Source $source): array
{
    	return $source->getTokens();
}

GromNaN replied:
Use php-cs-fixer to fix the coding style before committing changes, please.

owsPro replied:
I have it now.

Commit:
Update Environment.php

Rewritten: function compileSource

Commit:
Update Environment.php

Commit:
Update Environment.php

Commit:
Update Compiler.php

GromNaN commented on src/Compiler.php:
```diff
@@ -218,6 +227,6 @@ public function outdent(int $step = 1)
     public function getVarName(): string
     {
-        return sprintf('__internal_compile_%d', $this->varNameSalt++);
+        return \sprintf('__internal_compile_%d', $this->varNameSalt++);
```
The root namespace is added only for functions that have a special opcode compilation. This is done by php-cs-fixer. 
https://cs.symfony.com/doc/rules/function_notation/native_function_invocation.html

The function sprintf don't need to be root-namespaced.

owsPro replied:
Rule native_function_invocation¶
Adds a leading \ before the function call to speed up resolution.

Warning¶
Using this rule is risky¶
Risky if one of the functions is overridden.

Well, then you should never use rule native_function_invocation, as this could always be risky depending on the context, right?

xabbuh replied:
The point is not about the rule being risky but about the fact that the rule as applied here only prefixes some functions that can be optimized by the compiler with a backslash.

owsPro replied:
This is not objected to by the php cs fixer. This link https://stackoverflow.com/questions/55419673/php7-adding-a-slash-to-all-standard-php-functions-php-cs-fixer-rule shows that it is generally not that simple.

GromNaN commented on src/Environment.php:
```diff
@@ -40,7 +40,7 @@
 class Environment
 {
-    public const VERSION = '4.0.0-DEV';
+    public const VERSION = '4.0.0-DEV-modified-by-owsPro';
```
To be reverted.

owsPro replied:
It is now.

GromNaN reviewed (changes requested):
You are doing too many unrelated changes in this PR, it's not going to be something that can be merged. 
Please explain your goal so that we can help having a clean targeted diff.

smnandre commented on src/Compiler.php:
```diff
@@ -147,11 +141,11 @@ public function repr($value)
             $this->raw('array(');
             $first = true;
-            foreach ($value as $key => $v) {
+            \foreach ($value as $key => $v) {
```
Is phpcsfixer doing this ? oO 

I've never seen `\foreach` before

owsPro replied:
Nope, is corrected!

smnandre commented on src/Compiler.php:
```diff
@@ -167,27 +161,41 @@ public function repr($value)
+        return $this->debugInfo ?? [];
     }
+    
+    function sortData($data)
```
Where is it used ?

owsPro replied:
Has been removed again and should not appear at all!

smnandre commented on src/Environment.php:
```diff
@@ -798,9 +804,7 @@ public function mergeGlobals(array $context): array
-            if (!\array_key_exists($key, $context)) {
-                $context[$key] = $value;
-            }
+            $context[$key] = $value;
```
I think that would be a big BC break, no ?

stof commented on src/Environment.php:
```diff
@@ -136,7 +136,7 @@ public function enableDebug()
      * Disables debugging mode.
      */
-    public function disableDebug()
+    public function disableDebug(): void
```
Adding those return types should first be done in 3.x by adding `@return` annotations to warn child classes that they need to add it. Otherwise, we don't have the proper migration path.

owsPro replied:
Ok!

stof reviewed (commented):
With this PR that changes the formatting of the code not respecting the coding standards of the project, it is impossible to properly review it.
Thus, it does not even clearly explain its goal.

In its current state, my vote is to reject it.

Commit:
Update Compiler.php

Commit:
Update Environment.php

Commit:
Update Environment.php

smnandre commented on src/Environment.php:
```diff
@@ -514,17 +514,22 @@ public function compile(Node $node): string
+
+        return $content;
+    }
+    public function tokenize(Source $source): array
```
This looks like a c/c mistake ..  ?

fabpot commented:
@owsPro I don't see what you're trying to do here. Can you clarify your goal before investing more time? As of now, I think it's better to just close.

ghost commented:
If fabpot comes to the realisation that it is better to close here, it will probably be a wasted effort.

stof commented:
Well, this is the same than my comment: you don't explain what you are trying to do. That's the most important part when starting a PR.
